tests/ballerina-unit-test/src/test/java/org/ballerinalang/test/config/ConfigProcessorTest.java

```diff
@@ -0,0 +1,103 @@
+ *
+ * @since 0.982.0
+ */
+public class ConfigProcessorTest {
+
+    private Map<String, String> runtimeParams;
+    private String resourceRoot;
+    private String confFile;
+
+    @BeforeClass
+    public void setup() {
+        resourceRoot = new File(getClass().getProtectionDomain().getCodeSource().getLocation().getPath())
+                .getAbsolutePath();
+        confFile = Paths.get(resourceRoot, "datafiles", "config", "lookupenv.conf").toString();
+
+        runtimeParams = new HashMap<>();
+        runtimeParams.put("echo.http.host", "10.100.1.201");
+        runtimeParams.put("echo.http.port", "8082");
+    }
+
+    @Test
+    public void testProcessConfiguration() throws IOException {
+        BConfig bConfig = ConfigProcessor.processConfiguration(runtimeParams, confFile, null);
+        Map<String, Object> configs = bConfig.getConfigurations();
+
+        // Verify runtime params were correctly parsed and processed
+        Assert.assertTrue(bConfig.getConfigurations().keySet().containsAll(runtimeParams.keySet()));
+        Assert.assertTrue(bConfig.getConfigurations().values().containsAll(runtimeParams.values()));
+
+        // Verify config file was correctly parsed and env values were correctly looked up
+        Object obj = configs.get("hello.http.host");
+        Assert.assertTrue(obj instanceof String);
+        Assert.assertEquals(obj, "192.168.1.11");
+
+        obj = configs.get("hello.http.port");
+        Assert.assertTrue(obj instanceof Long);
+        Assert.assertEquals(obj, 5656L);
+
+        obj = configs.get("hello.cache.enabled");
+        Assert.assertTrue(obj instanceof Boolean);
+
+        obj = configs.get("hello.eviction.fac");
+        Assert.assertTrue(obj instanceof Double);
+        Assert.assertEquals(obj, 0.2333333D);
+    }
+
+    @Test(expectedExceptions = IllegalArgumentException.class,
+          expectedExceptionsMessageRegExp = ".*invalid int value received from environment to 'helloneg.http.port': " +
+                  "invalid port")
+    public void testEnvLookupNegative() throws IOException {
+        String negativeConf = Paths.get(resourceRoot, "datafiles", "config", "lookupenv_negative.conf").toString();
+        ConfigProcessor.processConfiguration(null, negativeConf, null);
+    }
+
+    @Test(expectedExceptions = IllegalArgumentException.class,
+          expectedExceptionsMessageRegExp = ".*invalid float value received from environment to 'helloneg.eviction" +
+                  ".fac': invalid eviction factor")
+    public void testEnvLookupNegative2() throws IOException {
+        String negativeConf = Paths.get(resourceRoot, "datafiles", "config", "lookupenv_negative2.conf").toString();
+        ConfigProcessor.processConfiguration(null, negativeConf, null);
+    }
+
+    // Not checking the error message in this because the order of the messages cannot be guaranteed.
```

Comment:
why the order is not guaranteed? Isn't the config file read in order?

Reply:
True, but the backing storage is a map. And in https://github.com/ballerina-platform/ballerina-lang/blob/bf6bda04e2519fff83400cae4d9c7d89dd4d1bdb/bvm/ballerina-config/src/main/java/org/ballerinalang/config/ConfigProcessor.java#L83 the entry set iterated to look up env vars.